state/src/item/metadata.rs

```diff
@@ -93,6 +95,14 @@ impl Metadata {
         self.params = Some(params);
     }
 
+    pub fn term_params(&self) -> Option<&CommonParams> {
```

Comment:
Why do you return `Option<&CommonParams>` instead of `&Option<CommonParams>`?

Reply:
The former has more advantages (importance order)
1. The former is more prevalent in Rust. More natural to look.
2. The former can avoid indirect memory reference when it just inspect whether it is some/none
3. They have same size (thanks to the null pointer optimization of Rust)
4. It is more flexible to construct, so it helps to maintain a stable API.
The former one can be created when there are a way to get following types of values from `self` in any mean
```rust
struct Foo {
	a: u64,
	b: Option<u64>,
}

impl Foo {
	fn c(&self) -> Option<&T> { ... }
	fn d(&self) -> &Option<T> { ... }

	fn option_ref(&self) -> Option<&T> {
		match random() % 4 {
			0 => Some(&self.a),
			1 => self.b.as_ref(),
			2 => self.c(),
			3 => self.d().as_ref(),
		}
	}

	fn ref_option(&self) -> &Option<T> {
		match random() % 4 {
			// 0 => Some(&self.a),
			1 => &self.b,
			// 2 => self.c(),
			3 => self.d().as_ref(),
		}
	}
}
```

Reply:
Since converting it to `Option<&T> is easier than the opposite, I thought returning `&Option<T>` is better because it can be used more widely.
But as you said, it's better to return `Option<&T>` because it's more convenient to construct. It'll be more stable even the structs changed.

By the way, would you explain more about `2. The former can avoid indirect memory reference when it just inspect whether it is some/none`?

Reply:
Sorry. I always confuse `reference` with `dereference`. `&Option<T>` is a reference to a value (containing a value) so we need to dereference to know whether it is `Some` or `None`, while we doesn't need to dereference since `Option<&T>` is just a value (containing a reference)

This one need deref whether to know it is `Some` or `None`.
```rust
pub fn is_some_ref_to_value(v: &Option<u32>) -> bool {
    v.is_some()
}
```
Here `[rdi]` means it reads memory whose address is `rdi`
```asm
cmp     dword ptr [rdi], 1
sete    al
ret
```
However, this function only performs register operations.
```rust
pub fn is_some_value_to_ref(v: Option<&u32>) -> bool {
    v.is_some()
}
```
```asm
test    rdi, rdi
setne   al
ret
```
See this disassembly result: https://godbolt.org/z/bbVa3N